### ISSUE TITLE
RangeError occurs with the released version of my app

### ISSUE THREAD

**Issue by Tolga-FNT** [simolus3/web3dart#47]
I have a method where I call my contract balance and decimals functions (see below):

```
case 'FNT':

        try {
        var abiCode = await DefaultAssetBundle.of(context).loadString('assets/json/fnt-abi.json');

        if ( abiCode != null ) {
          final EthereumAddress contractAddr = EthereumAddress.fromHex( '0x1f88badf3b56998673896572c8c1de846ab0e6c2' );
          var contract = DeployedContract ( ContractAbi.fromJson( abiCode, 'FinentToken' ), contractAddr );
          Credentials creds = EthPrivateKey.fromHex( crypto.privateKey );
          
          EthereumAddress address = await creds.extractAddress();
          // this is line 104 <-  CryptoHelper.loadBalance (package:finent_wallet/src/helpers/crypto-helper.dart:104)
          final balanceFunction = contract.function('balanceOf');
          final decimalsFunction = contract.function('decimals');
          
          final balance = await _client.call(
              contract: contract, function: balanceFunction, params: [address]);
          final decimals = await _client.call(
              contract: contract, function: decimalsFunction, params: []);
          
          BigInt bigBalance = BigInt.parse( balance.first.toString(), radix: 10 );
          BigInt bigDecimals = BigInt.parse( decimals.first.toString(), radix: 10 );

          cryptoBalances[crypto.symbol]['balance'] = bigBalance / BigInt.from(10).pow(bigDecimals.toInt());
          break;
        }
        cryptoBalances[crypto.symbol]['balance'] = 0;
        } catch( e ) {
          debugPrint( e.toString() );
          _showDialog( context, e.toString() );
        }
        break;
    }
```

And somehow it works when in development mode, but when creating a released version of the app, I get to see this error:

```
E/flutter (23672): [ERROR:flutter/lib/ui/ui_dart_state.cc(148)] Unhandled Exception: RangeError: Value not in range: 32
E/flutter (23672): #0      _rangeCheck (dart:typed_data-patch/typed_data_patch.dart:4445)
E/flutter (23672): #1      _ByteBuffer.asUint8List (dart:typed_data-patch/typed_data_patch.dart:1923)
E/flutter (23672): #2      _IntTypeBase.decode (package:web3dart/src/contracts/abi/integers.dart:23)
E/flutter (23672): #3      TupleType.decode (package:web3dart/src/contracts/abi/tuple.dart:102)
E/flutter (23672): #4      ContractFunction.decodeReturnValues (package:web3dart/src/contracts/abi/abi.dart:255)
E/flutter (23672): #5      Web3Client.call (package:web3dart/src/core/client.dart:303)
E/flutter (23672): <asynchronous suspension>
E/flutter (23672): #6      CryptoHelper.loadBalance (package:finent_wallet/src/helpers/crypto-helper.dart:104)
E/flutter (23672): <asynchronous suspension>
E/flutter (23672): #7      CryptoHelper.loadBalances (package:finent_wallet/src/helpers/crypto-helper.dart:59)
E/flutter (23672): <asynchronous suspension>
E/flutter (23672): #8      SplashState._navigateWhenUserExists (package:finent_wallet/src/screens/splash.dart:62)
E/flutter (23672): #9      _asyncThenWrapperHelper.<anonymous closure> (dart:async-patch/async_patch.dart:77)
E/flutter (23672): #10     _rootRunUnary (dart:async/zone.dart:1132)
E/flutter (23672): #11     _CustomZone.runUnary (dart:async/zone.dart:1029)
E/flutter (23672): #12     _FutureListener.handleValue (dart:async/future_impl.dart:126)
E/flutter (23672): #13     Future._propagateToListeners.handleValueCallback (dart:async/future_impl.dart:639)
E/flutter (23672): #14     Future._propagateToListeners (dart:async/future_impl.dart:668)
E/flutter (23672): #15     Future._complete (dart:async/future_impl.dart:473)
E/flutter (23672): #16     _SyncCompleter.complete (dart:async/future_impl.dart:51)
E/flutter (23672): #17     _AsyncAwaitCompleter.complete (dart:async-patch/async_patch.dart:28)
E/flutter (23672): #18     _completeOnAsyncReturn (dart:async-patch/async_patch.dart:294)
E/flutter (23672): #19     DatabaseHelper.getList (package:finent_wallet/src/core/database_helper.dart:0)
E/flutter (23672): #20     _asyncThenWrapperHelper.<anonymous closure> (dart:async-patch/async_patch.dart:77)
E/flutter (23672): #21     _rootRunUnary (dart:async/zone.dart:1132) 
E/flutter (23672): #22     _CustomZone.runUnary (dart:async/zone.dart:1029)
E/flutter (23672): #23     _FutureListener.handleValue (dart:async/future_impl.dart:126)
E/flutter (23672): #24     Future._propagateToListeners.handleValueCallback (dart:async/future_impl.dart:639)
E/flutter (23672): #25     Future._propagateToListeners (dart:async/future_impl.dart:668)
E/flutter (23672): #26     Future._complete (dart:async/future_impl.dart:473)
E/flutter (23672): #27     _SyncCompleter.complete (dart:async/future_impl.dart:51)
E/flutter (23672): #28     _AsyncAwaitCompleter.complete (dart:async-patch/async_patch.dart:28)
E/flutter (23672): #29     _completeOnAsyncReturn (dart:async-patch/async_patch.dart:294)
E/flutter (23672): #30     _SqfliteDatabaseBase&Object&SqfliteDatabaseMixin.txnSynchronized (package:sqflite/src/database_mixin.dart:0)
E/flutter (23672): #31     _asyncThenWrapperHelper.<anonymous closure> (dart:async-patch/async_patch.dart:77)
E/flutter (23672): #32     _rootRunUnary (dart:async/zone.dart:1132)
E/flutter (23672): #33     _CustomZone.runUnary (dart:async/zone.dart:1029)
E/flutter (23672): #34     _FutureListener.handleValue (dart:async/future_impl.dart:126)
E/flutter (23672): #35     Future._propagateToListeners.handleValueCallback (dart:async/future_impl.dart:639)
E/flutter (23672): #36     Future._propagateToListeners (dart:async/future_impl.dart:668)
E/flutter (23672): #37     Future._complete (dart:async/future_impl.dart:473)
E/flutter (23672): #38     _SyncCompleter.complete (dart:async/future_impl.dart:51)
E/flutter (23672): #39     _AsyncAwaitCompleter.complete (dart:async-patch/async_patch.dart:28)
E/flutter (23672): #40     _completeOnAsyncReturn (dart:async-patch/async_patch.dart:294)
E/flutter (23672): #41     BasicLock.synchronized (package:synchronized/src/basic_lock.dart:0)
E/flutter (23672): #42     _asyncThenWrapperHelper.<anonymous closure> (dart:async-patch/async_patch.dart:77)
E/flutter (23672): #43     _rootRunUnary (dart:async/zone.dart:1132)
E/flutter (23672): #44     _CustomZone.runUnary (dart:async/zone.dart:1029)
E/flutter (23672): #45     _FutureListener.handleValue (dart:async/future_impl.dart:126)
E/flutter (23672): #46     Future._propagateToListeners.handleValueCallback (dart:async/future_impl.dart:639)
E/flutter (23672): #47     Future._propagateToListeners (dart:async/future_impl.dart:668)
E/flutter (23672): #48     Future._complete (dart:async/future_impl.dart:473)
E/flutter (23672): #49     _SyncCompleter.complete (dart:async/future_impl.dart:51)
E/flutter (23672): #50     _AsyncAwaitCompleter.complete (dart:async-patch/async_patch.dart:28)
E/flutter (23672): #51     _completeOnAsyncReturn (dart:async-patch/async_patch.dart:294)
```

To create a released version, I use the following terminal commands:

flutter clean
flutter build apk --release

Any idea what is going wrong here?

**Comment by simolus3** [comment]
Huh that's interesting. Do help me investigate this, can you give me the ABI extract for the `balanceOf` and `decimals` function?
Also, do you have any chance to see the network requests? It would be interesting to see what was sent / returned when calling these functions.

**Comment by Tolga-FNT** [comment]
> Huh that's interesting. Do help me investigate this, can you give me the ABI extract for the `balanceOf` and `decimals` function?
> Also, do you have any chance to see the network requests? It would be interesting to see what was sent / returned when calling these functions.

```
[
	{
		"constant": true,
		"inputs": [],
		"name": "name",
		"outputs": [
			{
				"name": "",
				"type": "string"
			}
		],
		"payable": false,
		"stateMutability": "view",
		"type": "function"
	},
	{
		"constant": false,
		"inputs": [
			{
				"name": "spender",
				"type": "address"
			},
			{
				"name": "value",
				"type": "uint256"
			}
		],
		"name": "approve",
		"outputs": [
			{
				"name": "",
				"type": "bool"
			}
		],
		"payable": false,
		"stateMutability": "nonpayable",
		"type": "function"
	},
	{
		"constant": true,
		"inputs": [],
		"name": "totalSupply",
		"outputs": [
			{
				"name": "",
				"type": "uint256"
			}
		],
		"payable": false,
		"stateMutability": "view",
		"type": "function"
	},
	{
		"constant": true,
		"inputs": [],
		"name": "balanceOfZero",
		"outputs": [
			{
				"name": "",
				"type": "uint256"
			}
		],
		"payable": false,
		"stateMutability": "view",
		"type": "function"
	},
	{
		"constant": false,
		"inputs": [
			{
				"name": "_from",
				"type": "address"
			},
			{
				"name": "_to",
				"type": "address"
			},
			{
				"name": "_value",
				"type": "uint256"
			}
		],
		"name": "transferFrom",
		"outputs": [
			{
				"name": "",
				"type": "bool"
			}
		],
		"payable": false,
		"stateMutability": "nonpayable",
		"type": "function"
	},
	{
		"constant": true,
		"inputs": [],
		"name": "decimals",
		"outputs": [
			{
				"name": "",
				"type": "uint256"
			}
		],
		"payable": false,
		"stateMutability": "view",
		"type": "function"
	},
	{
		"constant": false,
		"inputs": [
			{
				"name": "spender",
				"type": "address"
			},
			{
				"name": "addedValue",
				"type": "uint256"
			}
		],
		"name": "increaseAllowance",
		"outputs": [
			{
				"name": "",
				"type": "bool"
			}
		],
		"payable": false,
		"stateMutability": "nonpayable",
		"type": "function"
	},
	{
		"constant": true,
		"inputs": [
			{
				"name": "owner",
				"type": "address"
			}
		],
		"name": "balanceOf",
		"outputs": [
			{
				"name": "",
				"type": "uint256"
			}
		],
		"payable": false,
		"stateMutability": "view",
		"type": "function"
	},
	{
		"constant": false,
		"inputs": [],
		"name": "acceptOwnership",
		"outputs": [],
		"payable": false,
		"stateMutability": "nonpayable",
		"type": "function"
	},
	{
		"constant": true,
		"inputs": [],
		"name": "getRescueAddr",
		"outputs": [
			{
				"name": "",
				"type": "address"
			}
		],
		"payable": false,
		"stateMutability": "view",
		"type": "function"
	},
	{
		"constant": false,
		"inputs": [
			{
				"name": "lostTokenContract",
				"type": "address"
			},
			{
				"name": "value",
				"type": "uint256"
			}
		],
		"name": "rescueLostTokenByThisTokenOwner",
		"outputs": [],
		"payable": false,
		"stateMutability": "nonpayable",
		"type": "function"
	},
	{
		"constant": false,
		"inputs": [
			{
				"name": "addr",
				"type": "address"
			}
		],
		"name": "setRescueAddr",
		"outputs": [],
		"payable": false,
		"stateMutability": "nonpayable",
		"type": "function"
	},
	{
		"constant": true,
		"inputs": [],
		"name": "owner",
		"outputs": [
			{
				"name": "",
				"type": "address"
			}
		],
		"payable": false,
		"stateMutability": "view",
		"type": "function"
	},
	{
		"constant": false,
		"inputs": [
			{
				"name": "addr",
				"type": "address"
			},
			{
				"name": "value",
				"type": "uint256"
			},
			{
				"name": "doFreeze",
				"type": "bool"
			}
		],
		"name": "distribute",
		"outputs": [],
		"payable": false,
		"stateMutability": "nonpayable",
		"type": "function"
	},
	{
		"constant": true,
		"inputs": [],
		"name": "symbol",
		"outputs": [
			{
				"name": "",
				"type": "string"
			}
		],
		"payable": false,
		"stateMutability": "view",
		"type": "function"
	},
	{
		"constant": false,
		"inputs": [
			{
				"name": "unfreezeDateTime",
				"type": "uint256"
			}
		],
		"name": "setUnfreezeDateTime",
		"outputs": [],
		"payable": false,
		"stateMutability": "nonpayable",
		"type": "function"
	},
	{
		"constant": false,
		"inputs": [
			{
				"name": "addr",
				"type": "address"
			},
			{
				"name": "value",
				"type": "uint256"
			}
		],
		"name": "burn",
		"outputs": [],
		"payable": false,
		"stateMutability": "nonpayable",
		"type": "function"
	},
	{
		"constant": false,
		"inputs": [
			{
				"name": "value",
				"type": "uint256"
			}
		],
		"name": "mint",
		"outputs": [],
		"payable": false,
		"stateMutability": "nonpayable",
		"type": "function"
	},
	{
		"constant": false,
		"inputs": [
			{
				"name": "spender",
				"type": "address"
			},
			{
				"name": "subtractedValue",
				"type": "uint256"
			}
		],
		"name": "decreaseAllowance",
		"outputs": [
			{
				"name": "",
				"type": "bool"
			}
		],
		"payable": false,
		"stateMutability": "nonpayable",
		"type": "function"
	},
	{
		"constant": false,
		"inputs": [
			{
				"name": "_to",
				"type": "address"
			},
			{
				"name": "_value",
				"type": "uint256"
			}
		],
		"name": "transfer",
		"outputs": [
			{
				"name": "",
				"type": "bool"
			}
		],
		"payable": false,
		"stateMutability": "nonpayable",
		"type": "function"
	},
	{
		"constant": true,
		"inputs": [],
		"name": "getUnfreezeDateTime",
		"outputs": [
			{
				"name": "",
				"type": "uint256"
			}
		],
		"payable": false,
		"stateMutability": "view",
		"type": "function"
	},
	{
		"constant": false,
		"inputs": [
			{
				"name": "value",
				"type": "uint256"
			}
		],
		"name": "burnFromOwner",
		"outputs": [],
		"payable": false,
		"stateMutability": "nonpayable",
		"type": "function"
	},
	{
		"constant": false,
		"inputs": [
			{
				"name": "target",
				"type": "address"
			},
			{
				"name": "doFreeze",
				"type": "bool"
			}
		],
		"name": "freeze",
		"outputs": [],
		"payable": false,
		"stateMutability": "nonpayable",
		"type": "function"
	},
	{
		"constant": true,
		"inputs": [],
		"name": "newOwner",
		"outputs": [
			{
				"name": "",
				"type": "address"
			}
		],
		"payable": false,
		"stateMutability": "view",
		"type": "function"
	},
	{
		"constant": true,
		"inputs": [
			{
				"name": "owner",
				"type": "address"
			},
			{
				"name": "spender",
				"type": "address"
			}
		],
		"name": "allowance",
		"outputs": [
			{
				"name": "",
				"type": "uint256"
			}
		],
		"payable": false,
		"stateMutability": "view",
		"type": "function"
	},
	{
		"constant": true,
		"inputs": [
			{
				"name": "target",
				"type": "address"
			}
		],
		"name": "isAccountFrozen",
		"outputs": [
			{
				"name": "",
				"type": "bool"
			}
		],
		"payable": false,
		"stateMutability": "view",
		"type": "function"
	},
	{
		"constant": true,
		"inputs": [],
		"name": "_unfreezeDateTime",
		"outputs": [
			{
				"name": "",
				"type": "uint256"
			}
		],
		"payable": false,
		"stateMutability": "view",
		"type": "function"
	},
	{
		"constant": false,
		"inputs": [
			{
				"name": "_newOwner",
				"type": "address"
			}
		],
		"name": "transferOwnership",
		"outputs": [],
		"payable": false,
		"stateMutability": "nonpayable",
		"type": "function"
	},
	{
		"constant": false,
		"inputs": [
			{
				"name": "lostTokenContract",
				"type": "address"
			},
			{
				"name": "value",
				"type": "uint256"
			}
		],
		"name": "rescueLostTokensByOwn",
		"outputs": [],
		"payable": false,
		"stateMutability": "nonpayable",
		"type": "function"
	},
	{
		"inputs": [],
		"payable": false,
		"stateMutability": "nonpayable",
		"type": "constructor"
	},
	{
		"payable": true,
		"stateMutability": "payable",
		"type": "fallback"
	},
	{
		"anonymous": false,
		"inputs": [
			{
				"indexed": false,
				"name": "",
				"type": "uint256"
			}
		],
		"name": "globalUnfreezeDatetimeModified",
		"type": "event"
	},
	{
		"anonymous": false,
		"inputs": [
			{
				"indexed": false,
				"name": "target",
				"type": "address"
			},
			{
				"indexed": false,
				"name": "frozen",
				"type": "bool"
			}
		],
		"name": "FreezeFunds",
		"type": "event"
	},
	{
		"anonymous": false,
		"inputs": [
			{
				"indexed": true,
				"name": "_from",
				"type": "address"
			},
			{
				"indexed": true,
				"name": "_to",
				"type": "address"
			}
		],
		"name": "OwnershipTransferred",
		"type": "event"
	},
	{
		"anonymous": false,
		"inputs": [
			{
				"indexed": true,
				"name": "from",
				"type": "address"
			},
			{
				"indexed": true,
				"name": "to",
				"type": "address"
			},
			{
				"indexed": false,
				"name": "value",
				"type": "uint256"
			}
		],
		"name": "Transfer",
		"type": "event"
	},
	{
		"anonymous": false,
		"inputs": [
			{
				"indexed": true,
				"name": "owner",
				"type": "address"
			},
			{
				"indexed": true,
				"name": "spender",
				"type": "address"
			},
			{
				"indexed": false,
				"name": "value",
				"type": "uint256"
			}
		],
		"name": "Approval",
		"type": "event"
	}
]
```
"Also, do you have any chance to see the network requests"
Uhm, how should I exactly do this? :)

**Comment by simolus3** [comment]
I guess the easiest way would be to temporarily add https://pub.dev/packages/http_logger, create a http client like
```dart
HttpWithMiddleware httpClient = HttpWithMiddleware.build(middlewares: [
  HttpLogger(logLevel: LogLevel.BODY),
]);
```
and pass that `httpClient` to the `Web3Client` constructor. The requests that contain `"method":"eth_call"` are interesting here.  Be aware that these potentially contain sensitive information (contract address, your ethereum address and the data sent and received from the contract). If you're on a dev or testing network that's fine, but if you're uncomfortable posting them publicly you could also sent them to me via oss@simonbinder.eu.

**Comment by Tolga-FNT** [comment]
I can't use both web3dart, http and http_logger because of the version issues. See the logs below:

```
So, because fw depends on both http_logger ^1.0.0 and http ^0.12.0, version solving failed.
pub get failed (1)
exit code 1
[fw] flutter packages get
Running "flutter packages get" in fw...              
Because web3dart >=0.4.2 depends on http ^0.12.0 and fw depends on http ^0.11.3+16, web3dart >=0.4.2 is forbidden.

So, because fw depends on web3dart ^1.0.0-rc.2, version solving failed.
pub get failed (1)
exit code 1
[fw] flutter packages get
Running "flutter packages get" in fw...              
Because http_logger >=1.0.0 depends on http_middleware ^1.0.0 which depends on http ^0.11.3+16, http_logger >=1.0.0 requires http ^0.11.3+16.

So, because fw depends on both http_logger ^1.0.0 and http ^0.12.0+2, version solving failed.
pub get failed (1)
exit code 1
```

**Comment by simolus3** [comment]
Sorry, forget about `http_logger`. 
You can write one yourself with
```dart
import 'package:http/http.dart';
import 'package:web3dart/web3dart.dart';

class LoggingClient extends BaseClient {
  final Client _inner;

  LoggingClient(this._inner);

  @override
  Future<StreamedResponse> send(BaseRequest request) async {
    if (request is Request) {
      print('sending ${request.url} with ${request.body}');
    } else {
      print('sending ${request.url}');
    }

    final response = await _inner.send(request);
    final read = await Response.fromStream(response);

    print('response:\n${read.body}');

    return StreamedResponse(
        Stream.fromIterable([read.bodyBytes]), response.statusCode);
  }
}
```
Then, you can create the client with `Web3Client(rpcUrl, LoggingClient(Client()))`, it should log outgoing requests and their responses.

**Comment by simolus3** [comment]
Thanks for the email. It appears like this was a Dart (compiler?) bug that is fixed in the beta channel of Flutter.

**Comment by thelazyindian** [comment]
Well, I am having the same issue 
How to fix?

**Comment by simolus3** [comment]
@thelazyindian Can you open another issue with a simple repro that behaves differently on debug and release builds? It would also be helpful to know which exact Flutter version you're using. Thanks.

**Comment by thelazyindian** [comment]
Nvm. This error also comes when the contract address is wrong. I messed the contract address earlier but fixed it now and so the error is also fixed. Sorry for incovenience.
